server/src/main/java/io/crate/execution/engine/collect/DocValuesGroupByOptimizedIterator.java

```diff
@@ -0,0 +1,428 @@
+            MappedFieldType keyFieldType = fieldTypeLookup.get(keyRef.column().fqn());
+            if (keyFieldType == null || !keyFieldType.hasDocValues()) {
+                return null;
+            } else {
+                columnKeyRefs.add(keyColumnRef);
+            }
+        }
+
+        //noinspection rawtypes
+        List<DocValueAggregator> aggregators = DocValuesAggregates.createAggregators(
+            functions,
+            groupProjection.values(),
+            fieldTypeLookup,
+            collectPhase.toCollect(),
+            collectTask.txnCtx().sessionSettings().searchPath()
+        );
+        if (aggregators == null) {
+            return null;
+        }
+
+        if (Symbols.containsColumn(collectPhase.toCollect(), DocSysColumns.SCORE)
+            || Symbols.containsColumn(collectPhase.where(), DocSysColumns.SCORE)) {
+            return null;
+        }
+
+        ShardId shardId = indexShard.shardId();
+        SharedShardContext sharedShardContext = collectTask.sharedShardContexts().getOrCreateContext(shardId);
+        Engine.Searcher searcher = sharedShardContext.acquireSearcher(formatSource(collectPhase));
+
+        try {
+            QueryShardContext queryShardContext = sharedShardContext.indexService().newQueryShardContext();
+            collectTask.addSearcher(sharedShardContext.readerId(), searcher);
+
+            InputFactory.Context<? extends LuceneCollectorExpression<?>> docCtx
+                = docInputFactory.getCtx(collectTask.txnCtx());
+            docCtx.add(columnKeyRefs);
+            List<? extends LuceneCollectorExpression<?>> keyExpressions = docCtx.expressions();
+
+            LuceneQueryBuilder.Context queryContext = luceneQueryBuilder.convert(
+                collectPhase.where(),
+                collectTask.txnCtx(),
+                indexShard.mapperService(),
+                indexShard.shardId().getIndexName(),
+                queryShardContext,
+                table,
+                sharedShardContext.indexService().cache()
+            );
+
+            var ramAccounting = collectTask.getRamAccounting();
+            if (columnKeyRefs.size() == 1) {
+                return GroupByIterator.forSingleKey(
+                    aggregators,
+                    searcher,
+                    columnKeyRefs.get(0),
+                    keyExpressions,
+                    ramAccounting,
+                    queryContext.query(),
+                    new CollectorContext(sharedShardContext.readerId())
+                );
+            } else {
+                return GroupByIterator.forManyKeys(
+                    aggregators,
+                    searcher,
+                    columnKeyRefs,
+                    keyExpressions,
+                    ramAccounting,
+                    queryContext.query(),
+                    new CollectorContext(sharedShardContext.readerId())
+                );
+            }
+        } catch (Throwable t) {
+            searcher.close();
```

Comment:
I wonder if we have a potential double-close issue here. I think after the `collectTask.addSearcher(sharedShardContext.readerId(), searcher);` call the searcher will be closed as part of the `collectTask`.

This is probably also the case in the other places where we have the same setup pattern - so may not be exclusively to this PR now and we can also follow up on it later.

Reply:
👍 added it into the follow-up list. i'll check it later.